File: Stack/SpecialStack/java/code.java
import java.util.Stack;

class Stack {

	class SpecialStack {
		Stack<Integer> s;
		int min;

		SpecialStack() {
			s = new Stack<Integer>();
			min=-1;
		}

		void push(int x) {
			s.push(x);
			if(min==-1) {
				min=x;
			}
			else {
				if(min>x) {
					min=x;
				}
			}
		}

		void pop() {
			s.pop();
		}

		bool isEmpty() {
			return s.isEmpty();
		}

		int minVal() {
			return min;
		}

		int top() {
			return s.top();
		}
	}

	public static void main(String[] args) {

		SpecialStack s;

		s.push(1);
		s.push(2);
		s.push(3);
		s.push(4);

		System.out.println(s.min());
		System.out.println(s.top());
		
	}
}

File: Stack/SpecialStack/cpp/code.cpp
#include <bits/stdc++.h>
using namespace std;

class SpecialStack {
private:
	stack<int> s;
	stack<int> s_min;

public:
	void push(int x) {
		s.push(x);
		if(s_min.empty()) {
			s_min.push(x);
		}
		else {
			if(s_min.top()>x) {
				s_min.push(x);
			}
		}
	}

	int top() {
		if(s.empty()) {
			exit(0);
		}
		return s.top();
	}

	void pop() {
		if(s.empty()) {
			exit(0);
		}
		s.pop();
	}

	bool isEmpty() {
		if(s.empty()) {
			return true;
		}
		return false;
	}

	int min() {
		if(s_min.empty()) {
			return -1;
		}
		else {
			return s_min.top();
		}
	}
};

int main() {

	SpecialStack s;

	s.push(1);
	s.push(2);
	s.push(3);
	s.push(4);

	cout << s.min() << endl;
	cout << s.top() << endl;

	return 0;
}